cmd/aws-vpc-cni-init/main.go

```diff
@@ -0,0 +1,195 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// The aws-node initialization
+package main
+
+import (
+	"os"
+
+	"github.com/aws/amazon-vpc-cni-k8s/utils/cp"
+	"github.com/aws/amazon-vpc-cni-k8s/utils/imds"
+	"github.com/aws/amazon-vpc-cni-k8s/utils/sysctl"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+)
+
+const (
+	defaultHostCNIBinPath = "/host/opt/cni/bin"
+	vpcCniInitDonePath    = "/vpc-cni-init/done"
+	metadataLocalIP       = "local-ipv4"
+	metadataMAC           = "mac"
+
+	envDisableIPv4TcpEarlyDemux = "DISABLE_TCP_EARLY_DEMUX"
+	envEnableIPv6               = "ENABLE_IPv6"
+	envHostCniBinPath           = "HOST_CNI_BIN_PATH"
+)
+
+func getEnv(env, defaultVal string) string {
+	if val, ok := os.LookupEnv(env); ok {
+		return val
+	}
+	return defaultVal
+}
+
+func getNodePrimaryIF() (string, error) {
+	var primaryIF string
+	primaryMAC, err := imds.GetMetaData("mac")
+	if err != nil {
+		return primaryIF, errors.Wrap(err, "Failed to get primary MAC from IMDS")
+	}
+	log.Infof("Found primaryMAC %s", primaryMAC)
+
+	links, err := netlink.LinkList()
+	if err != nil {
+		return primaryIF, errors.Wrap(err, "Failed to list links")
+	}
+	for _, link := range links {
+		if link.Attrs().HardwareAddr.String() == primaryMAC {
+			primaryIF = link.Attrs().Name
+			break
+		}
+	}
+
+	if primaryIF == "" {
+		return primaryIF, errors.Wrap(err, "Failed to retrieve primary IF")
+	}
+	return primaryIF, nil
+}
+
+func configureSystemParams(sysctlUtil sysctl.Interface, primaryIF string) error {
+	var err error
+	// Configure rp_filter in loose mode
+	entry := "net/ipv4/conf/" + primaryIF + "/rp_filter"
+	err = sysctlUtil.Set(entry, 2)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to set rp_filter for %s", primaryIF)
+	}
+	val, _ := sysctlUtil.Get(entry)
+	log.Infof("Updated %s to %d", entry, val)
+
+	// Enable or disable TCP early demux based on environment variable
+	// Note that older kernels may not support tcp_early_demux, so we must first check that it exists.
+	entry = "net/ipv4/tcp_early_demux"
+	if _, err := sysctlUtil.Get(entry); err != nil {
```

Comment:
shouldn't this be `err == nil`?

Reply:
If the path to `tcp_early_demux` does not exist, we return error. If an error is returned, we do not want to write to `tcp_early_demux`

Reply:
Oh, nvm, you are right here. I was not being thorough in testing after this change

Reply:
fixed in latest revision